### DISCUSSION TITLE
How to fetch the latest code version of an asset?

### DISCUSSION THREAD

**Discussion by clairelin135** [dagster-io/dagster#17426] [Q&A] [answered]
How do I fetch the latest code version of an asset so I can determine if the code version has changed?

**Comment by clairelin135** [comment, marked as answer]
You can call:
```python
context.instance.get_latest_materialization_code_versions(asset_keys)
```
to get the latest materialization code version per asset. 

From there, you can compare against the current code version of the asset.

**Comment by paulBurnsUpside** [comment]
Is there a way to automatically define a code_version for assets created through @asset instead of needing to manually define the code_version?

**Reply by garethbrickman** [reply]
You can use a function to generate a hash of the asset's source code, which can then be used as the code_version. This approach allows you to automatically version your assets based on their source code, which can be particularly useful for tracking changes and avoiding redundant computation.

Here's an example [from the documentation](https://docs.dagster.io/guides/dagster/asset-versioning-and-caching) that demonstrates how to use a function to automatically generate a code_version:

```python
from dagster import asset
import inspect
import hashlib

def get_function_hash(func):
    source_code = inspect.getsource(func)
    return hashlib.md5(source_code.encode()).hexdigest()

@asset(code_version=get_function_hash(my_asset))
def my_asset():
    # asset definition
    ...
```

In this example, the `get_function_hash` function takes a Python function as an argument, retrieves its source code, and then computes an MD5 hash of the source code. This hash is then used as the code_version for the my_asset asset.

It's important to note that this approach has some limitations. For instance, it only considers the source code of the function itself and does not account for changes in dependencies or external modules that the function may use. Additionally, this method may not be suitable for all use cases, and you should consider whether it fits your specific needs for asset versioning.

There is [an open feature request](https://github.com/dagster-io/dagster/issues/15242) discussing the possibility of implementing an "automatic source code versioning" feature at the asset or global level, which would further streamline this process.

**Reply by DennisSchwartz** [reply]
This unfortunately doesn't work since `my_asset` is referenced before it's been defined. I was wondering about that but thought maybe decorators are somehow special. Unless I'm doing something completely wrong here.

**Reply by garethbrickman** [reply]
Ah you're right I missed that reference before assignment, my apologies. I think to do that then would require [the implementation that is proposed in this feature request ](https://github.com/dagster-io/dagster/issues/15242#issuecomment-1634578494)
